### PR TITLE
JIT: Remove "promoted struct death vars" map

### PR DESCRIPTION
**Pull request by jakobbotsch**
Repurpose the multireg var death flags as general purpose "field death" flags instead. Update the TreeLifeUpdater to take this into account. Update dumping of promoted locals to always display last uses of fields. 

Couldn't reopen my old PR because of force pushing.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
This is blocked on #80611 for now.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @BruceForstall 

[Only a couple of diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=142223&view=ms.vss-build-web.run-extensions-tab). The new logic allows marking dead fields as dead when a promoted struct is being born, which the old logic did not.
This rarely allows some new copy props (only rarely because the dead assignments would usually be eliminated anyway, except when they are to a promoted struct that has significant padding and holes).

File: src/coreclr/jit/morphblock.cpp [hunk 1]
```diff
@@ -1272,7 +1272,12 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
             dstFld = m_comp->gtNewLclvNode(dstFieldLclNum, m_comp->lvaGetDesc(dstFieldLclNum)->TypeGet());
 
             // If it had been labeled a "USEASG", assignments to the individual promoted fields are not.
-            dstFld->gtFlags |= m_dstLclNode->gtFlags & ~(GTF_NODE_MASK | GTF_VAR_USEASG);
+            dstFld->gtFlags |= m_dstLclNode->gtFlags & ~(GTF_NODE_MASK | GTF_VAR_USEASG | GTF_VAR_DEATH_MASK);
+
+            if (m_dstLclNode->IsLastUse(i))
+            {
+                dstFld->gtFlags |= GTF_VAR_DEATH;
```
Comment (jakobbotsch) [inline src/coreclr/jit/morphblock.cpp:1279]: We can avoid creating IR for these stores at all (#80498). I'll handle that in a follow-up.

File: src/coreclr/jit/forwardsub.cpp [hunk 2]
```diff
@@ -360,14 +360,17 @@ class ForwardSubVisitor final : public GenTreeVisitor<ForwardSubVisitor>
...
 
-        if ((lcl->gtFlags & GTF_VAR_DEATH) == 0)
+        LclVarDsc* dsc = m_compiler->lvaGetDesc(lcl);
+
+        if (dsc->lvPromoted)
         {
-            return false;
+            assert(!dsc->lvTracked);
+            GenTreeFlags allFieldsDying =
+                static_cast<GenTreeFlags>(((1 << dsc->lvFieldCnt) - 1) << FIELD_LAST_USE_SHIFT);
```
[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/forwardsub.cpp:369]: Looks like this code pattern exists elsewhere, as well (e.g., fgMorphExpandImplicitByRefArg)

**Review by BruceForstall — APPROVED** [review]
LGTM. One minor suggestion. It's awesome to get rid of this map!

File: src/coreclr/jit/forwardsub.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/forwardsub.cpp]: I added `LclVarDsc::AllFieldDeathFlags()` (for promoted locals) and `LclVarDsc::FullDeathFlags()` (for any local)

**Comment by BruceForstall** [comment]
I like the cleanup with the new functions, but it looks like your new assert is hitting:
```
Assertion failed 'lvPromoted && (lvFieldCnt <= 4)' in 'System.Decimal:Ceiling(System.Decimal):System.Decimal' during 'Morph - Global' (IL size 36; hash 0xb0841e47; FullOpts)
```

btw, any thoughts on what should happen when (if?) we generalize struct promotion to more than 4 fields, or not all the fields?

**Comment by jakobbotsch** [comment]
> I like the cleanup with the new functions, but it looks like your new assert is hitting:
> 
> ```
> Assertion failed 'lvPromoted && (lvFieldCnt <= 4)' in 'System.Decimal:Ceiling(System.Decimal):System.Decimal' during 'Morph - Global' (IL size 36; hash 0xb0841e47; FullOpts)
> ```

Yep, I'll take a look next week. Maybe some of the state is desynced for promoted implicit byrefs while they are being morphed.

> btw, any thoughts on what should happen when (if?) we generalize struct promotion to more than 4 fields, or not all the fields?

The plan is that the generalized promotion pass works only (mostly) by inserting new IR and does not need the special treatment in downstream passes that promoted structs and their fields have today.

**Comment by jakobbotsch** [comment]
> Maybe some of the state is desynced for promoted implicit byrefs while they are being morphed.

Nope, the problem is that the we can get here for fields of promoted implicit byrefs too. The logic was technically incorrect there (but ok for the optimization we do with the information). Fixed it, will probably result in a few minor diffs.

**Pull request merged**